### PR TITLE
Sim Control Flow Feature Addition

### PR DESCRIPTION
**Pull request by nruof**
Class that allows a user to define a list of function criteria to be evaluated over a set of diagnostic steps in a simulation.

Example:

steps_per_loop = 10
        while (sim_control.check_criteria()):
            sim.step(steps_per_loop)

### CODE REVIEW

File: mewarpx/mewarpx/sim_control.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+        criteria_args: list of kwargs for each criteria
+        """
+
+        self.crit_list = criteria
+        self.crit_args_list = criteria_args
+        self.max_loops = max_loops
+        self.check_increment = 0
+
+    def add_checker(self, criteria=None, crit_args=None):
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/sim_control.py]: Not a big deal, but there doesn't seem to be a use case for where `criteria = None`, so `criteria` shouldn't have a default value of `None`.

File: mewarpx/mewarpx/sim_control.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        self.crit_list.append(criteria)
+        if crit_args:
+            self.crit_args_list.append(crit_args)
+        else:
+            self.crit_args_list.append({})
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/sim_control.py]: Here you don't need the if else statement/branching logic if you set the default value of `crit_args` to `crit_args={}`

File: mewarpx/mewarpx/sim_control.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    def check_criteria(self):
+        """
+        Evaluates each criteria in the crit_list
+        returns True is all criteria are True otherwise returns False
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/sim_control.py]:
```suggestion
        returns True if all criteria are True otherwise returns False
```
A small typo

**Review by KZhu-ME — COMMENTED** [review]
Looks good, just a have a few minor suggestions.

File: mewarpx/mewarpx/sim_control.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+# Get module-level logger
+logger = logging.getLogger(__name__)
+
+class SimControl:
+    """
+    Evaluate whether to continue or terminate the simulation based on a
+    set of user defined functions and criteria
+
+    """
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/sim_control.py]: This class should definitely allow for the scenario where `max_loops` is the only argument specified. So I'd suggest giving default values of `None` for `criteria` and `criteria_args` and not setting a default value for max_loops. Then add check in the __init__ function that max_loops >= 1.

File: mewarpx/mewarpx/sim_control.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import numpy as np
+import psutil
+
+from pywarpx import _libwarpx
+from mewarpx.mwxrun import mwxrun
+
+# Get module-level logger
+logger = logging.getLogger(__name__)
+
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py:6]:
```suggestion


```

File: mewarpx/mewarpx/sim_control.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                    print(("Termination due to criteria {}").format(func.__name__))
+                    return keep_going
+            self.check_increment += 1
+            if self.max_loops:
+                if self.check_increment >= self.max_loops:
+                    keep_going = False
+            return keep_going
+        else:
+            self.check_increment += 1
+            return True
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/sim_control.py]: With the forcing of a `max_loops` value from above I'd change this function to something like
```suggestion
        self.check_increment += 1
        if self.check_increment >= self.max_loops:
            print('Max loops reached.')
            return False
            
        for i, func in enumerate(self.crit_list):
            if not self.crit_list[i](**self.crit_args_list[i]):
                print(("Termination due to criteria {}").format(func.__name__))
                return False
```
But actually even better than using `max_loops` argument would be to use a `max_steps` argument and then calling the mewarpx function that returns the simulation step number when evaluating whether to terminate.

File: mewarpx/mewarpx/sim_control.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+""" Control code used to terminate simulation based on
+a set of user defined criteria
+"""
+
+import logging
+import time
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py]: Remove unused import

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
I left some comments on suggested changes, please let me know if anything is unclear.

File: mewarpx/mewarpx/sim_control.py (hunk 7 above)
Comment (peterscherpelz) [reply mewarpx/mewarpx/sim_control.py]: and numpy, psutil, _libwarpx, mwxrun - or at least comment out if they'll be later used as you expand this.

File: mewarpx/mewarpx/sim_control.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+class SimControl:
+    """
+    Evaluate whether to continue or terminate the simulation based on a
+    set of user defined functions and criteria
+
+    """
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
+        """
+        Generate and install functions to perform after a step #.
+        Arguments:
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py:17]: Newline before Arguments will make autodoc happier. Then indent criteria and criteria_args by 4 spaces relative to arguments.
Also add max_loops and kwargs to Arguments list
You should either say 'read check_criteria' documentation below, or move info about criteria up here - eg when do we want it to return True vs False.

File: mewarpx/mewarpx/sim_control.py (hunk 4 above)
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py]: I would not separate criteria and criteria_args into separate lists. (I don't like the format of having to match up correct criteria with their args.)
A nice format might be a list of criteria, and each criteria can either be a function or a tuple of `(function, kwargs)`.
```
if callable(criteria[i]):
   self.crit_list.append(criteria[i])
   self.crit_args_list.append({})
else:
   self.crit_list.append(criteria[i][0])
   self.crit_args_list.append(criteria[i][1])
```

File: mewarpx/mewarpx/sim_control.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        criteria: list of user defined functions that return True or False value
+
+        criteria_args: list of kwargs for each criteria
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py]: dict, not list

File: mewarpx/mewarpx/sim_control.py (hunk 2 above)
Comment (peterscherpelz) [reply mewarpx/mewarpx/sim_control.py]: You actually don't want to do this as a general rule.
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html or a more verbose explanation 
https://towardsdatascience.com/python-pitfall-mutable-default-arguments-9385e8265422

File: mewarpx/mewarpx/sim_control.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        steps_per_loop = 10
+        while (sim_control.check_criteria()):
+            sim.step(steps_per_loop)
+
+        """
+        if self.check_increment > 0:
+            keep_going = False
+            for i, func in enumerate(self.crit_list):
+                keep_going = self.crit_list[i](**self.crit_args_list[i])
+                if not keep_going:
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/sim_control.py]: I would not short-circuit these function evaluations if one returns false. Sometimes we'll want to install functions that both handle diagnostics and tell something to terminate. It'll probably be most straightforward if those all still run on the last time through things.

Comment (peterscherpelz) [reply mewarpx/mewarpx/sim_control.py]: You can keep the same print info by starting with an empty string, then add to it for each criteria that returns False.

**Review by peterscherpelz — CHANGES_REQUESTED** [review]
Lots of comments, mostly for writing clean code in the long-term. Thanks!

**Review by KZhu-ME — APPROVED** [review]
Aside from the changes Peter suggested, everything else looks good to me!

**Review by EtherealMC-Bit — APPROVED** [review]
Cool stuff

**Comment by nruof** [comment]
I tested out the changes.  eval_max_steps terminates the simulation properly a custom criteria function also terminates properly.  Changes are good.

**Review by roelof-groenewald — APPROVED** [review]
Looks good to me.

**Pull request merged**